### PR TITLE
Optimize t5 encoder in beam search

### PR DESCRIPTION
**Pull request by wangyems**
**Description**: Describe your changes.
Run EncoderDecoderInit with original inputs and expand the output.
zcode(cpu) inference time 17s -> 11s

**Motivation and Context**
- Why is this change required? What problem does it solve?
- If it fixes an open issue, please link to the issue here.

### CODE REVIEW

**Comment by tianleiwu** [comment]
Does CUDA implementation need any change? Please run test\python\transformers\test_beam_search.py in CPU and GPU to verify the change.

File: onnxruntime/contrib_ops/cpu/transformers/beam_search_impl_base.h [hunk 1]
```diff
@@ -109,6 +109,21 @@ struct BeamSearchCpuState : public IBeamSearchCpuState {
...
+  void SetSequence(gsl::span<const int32_t> input_ids_in_cpu,
+                   size_t batch_beam_size,
+                   int beam_size,
+                   int max_length,
+                   int sequence_length) {
+    gsl::span<int32_t> sequences_0 = sequences_space;
+    for (size_t i = 0; i < batch_beam_size; i++) {
+      for (int j = 0; j < sequence_length; j++) {
+        const size_t index = SafeInt<gsl::index>(i) * max_length + j;
+        sequences_0[index] = input_ids_in_cpu[SafeInt<gsl::index>(i) * sequence_length * beam_size + j];
```
[the remark this answers is not in the document]
Comment (wangyems) [reply onnxruntime/contrib_ops/cpu/transformers/beam_search_impl_base.h:122]: good catch!

File: onnxruntime/contrib_ops/cpu/transformers/subgraph_t5_decoder.cc [hunk 2]
```diff
@@ -156,9 +156,15 @@ Status T5DecoderSubgraph::CreateInitialFeeds(
   // of encoder.
   // When first_past_input_index_ == 2, the past states are copied from the second output of encoder.
   for (size_t j = 4 - first_past_input_index_; j < encoder_fetches.size(); j++) {
-    OrtValue expanded_cache;
-    BeamSearchCpuDeviceHelper::ExpandCaches<float>(encoder_fetches[j], num_beam, allocator, expanded_cache);
-    decoder_feeds.push_back(expanded_cache);
+    if (j == 1) {
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cpu/transformers/subgraph_t5_decoder.cc:159]: hidden_states is optional, need add a check that there is hidden state.

File: onnxruntime/contrib_ops/cpu/transformers/beam_search_device_helper.cc [hunk 3]
```diff
@@ -237,33 +267,30 @@ Status ProcessLogits(const OrtValue& logits,                                 //
   //    next_token_logits = logits[:, -1, :], and the result shape is (batch_size * num_beams, vocab_size)
   // When input_length == 1, use logits directly in SoftmaxCPU below so it only need for input_length > 1.
   gsl::span<T>& next_token_logits = beam_state->next_token_logits;
-  if (input_length > 1) {
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cpu/transformers/beam_search_device_helper.cc:208]: When input_length==1 and beam_batch_size == batch_beam_size, there is only one token so we can use logits as next_token_logits.

File: onnxruntime/contrib_ops/cpu/transformers/subgraph_t5_decoder.cc (hunk 2 above)
Comment (wangyems) [reply onnxruntime/contrib_ops/cpu/transformers/subgraph_t5_decoder.cc:170]: check added

File: onnxruntime/contrib_ops/cpu/transformers/beam_search_device_helper.cc (hunk 3 above)
Comment (wangyems) [reply onnxruntime/contrib_ops/cpu/transformers/beam_search_device_helper.cc:208]: good catch!

**Comment by wangyems** [comment]
> Does CUDA implementation need any change? Please run test\python\transformers\test_beam_search.py in CPU and GPU to verify the change.

all tests passed

File: onnxruntime/contrib_ops/cpu/transformers/beam_search_device_helper.cc [hunk 4]
```diff
@@ -60,6 +60,47 @@ void ExpandInputs(const OrtValue& input, int num_beams, AllocatorPtr allocator,
   }
 }
 
+template <typename T>
+Status ExpandBuffer(void* stream,
+                  const OrtValue& input,
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cpu/transformers/beam_search_device_helper.cc:65]: alignment

Comment (wangyems) [reply onnxruntime/contrib_ops/cpu/transformers/beam_search_device_helper.cc]: fixed

File: onnxruntime/contrib_ops/cpu/transformers/beam_search_device_helper.cc [hunk 5]
```diff
@@ -200,37 +241,43 @@ Status ProcessLogits(const OrtValue& logits,                                 //
   const TensorShape& logits_shape = logits.Get<Tensor>().Shape();
   ORT_ENFORCE(logits_shape.NumDimensions() == 3);
   auto input_length = logits_shape[1];
+  auto beam_batch_size = logits_shape[0];
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cpu/transformers/beam_search_device_helper.cc:244]: nit: it might be better to name as logits_batch_size

File: onnxruntime/contrib_ops/cpu/transformers/beam_search_device_helper.cc [hunk 6]
```diff
@@ -200,37 +241,43 @@ Status ProcessLogits(const OrtValue& logits,                                 //
...
+      }
     }
   }
 
 #ifdef DEBUG_BEAM_SEARCH
   dumper->Print("logits", logits);
-  if (input_length > 1) {
-    dumper->Print("next_token_logits", next_token_logits.data(), batch_size, num_beams, vocab_size);
-  }
+  dumper->Print("next_token_logits", next_token_logits.data(), batch_size, num_beams, vocab_size);
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cpu/transformers/beam_search_device_helper.cc:268]: nit: could we add condition if (input_length > 1 || beam_batch_size == batch_size), otherwise next_token_logits might not have value for print.

**Review by tianleiwu — DISMISSED** [review]

File: onnxruntime/contrib_ops/cpu/transformers/beam_search_device_helper.cc (hunk 6 above)
Comment (wangyems) [reply onnxruntime/contrib_ops/cpu/transformers/beam_search_device_helper.cc:268]: done

File: onnxruntime/contrib_ops/cpu/transformers/beam_search_device_helper.cc [hunk 7]
```diff
@@ -60,6 +60,47 @@ void ExpandInputs(const OrtValue& input, int num_beams, AllocatorPtr allocator,
   }
 }
 
+template <typename T>
```
[the remark this answers is not in the document]
Comment (wangyems) [reply onnxruntime/contrib_ops/cpu/transformers/beam_search_device_helper.cc:63]: added a TODO

**Review by tianleiwu — APPROVED** [review]
